patches/minecraft/net/minecraft/item/crafting/RecipeManager.java.patch

```diff
@@ -18,3 +29,40 @@
              } else {
                 this.func_199509_a(RecipeSerializers.func_199572_a(resourcelocation1, jsonobject));
              }
+@@ -67,9 +72,12 @@
+          throw new IllegalStateException("Duplicate recipe ignored with ID " + p_199509_1_.func_199560_c());
+       } else {
+          this.field_199522_d.put(p_199509_1_.func_199560_c(), p_199509_1_);
++         if(!p_199509_1_.getType().getBaseClass().isAssignableFrom(p_199509_1_.getClass())) throw new IllegalStateException(String.format("Recipe type %s is not valid for class %s", p_199509_1_.getType().getBaseClass().getName(), p_199509_1_.getClass().getName()));
++         ((java.util.List) getRecipes(p_199509_1_.getType())).add(p_199509_1_);
```

Comment:
Is there a reason we are forcing the base type? Is there a area in vanilla that makes that requirement?

Reply:
Yes.  TileEntityFurnaces casts IRecipe to FurnaceRecipe unchecked, by grabbing all recipes of type furnace.  This type enforcement allows for extra data (like xp, cook time, etc.) to be stored on the recipe.